src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs

```diff
@@ -478,14 +495,25 @@ matchNonVarToPattern
             m
         )
         (OrOfPredicateSubstitution level variable)
-matchNonVarToPattern tools substitutionSimplifier first second =
+matchNonVarToPattern tools substitutionSimplifier first second
+  = do
     -- TODO(virgil): For simplification axioms this would need to return bottom!
-    MaybeT $ lift $ do -- MonadCounter
+    finalResult <- MaybeT $ lift $ do -- MonadCounter
         (result, _proof) <-
             Equals.makeEvaluateTermsToPredicateSubstitution
                 tools substitutionSimplifier first second
         (return . return)
             (fmap secondVariablesSubstitutionToPredicate result)
+    let
+        firstVariables = freePureVariables first
+        resultVariables =
+            foldMap
+                (freeVariablesUnderArbitrarySymbols tools)
+                (OrOfExpandedPattern.extractPatterns finalResult)
+    if null (Set.intersection firstVariables resultVariables)
```

Comment:
Could you add a comment here explaining what the logic is? I'm struggling to understand the big picture here.

Reply:
Thanks, I should have done that before sending the code review. Is it easy enough to understand now?

Reply:
:+1